### ISSUE TITLE
Plugins: Input and output ports should set different default values for nChannels and nSampleRate

### ISSUE THREAD

**Issue by CapOM** [tizonia/tizonia-openmax-il#399]
Hi Juan, I am trying to enable OMX.Aratelia.audio_decoder.aac in gst-omx.

For audio and video decoders in gst-omx, it always waits for a OMX_EventPortSettingsChanged on the output port. If it never comes then gst-omx waits forever doing nothing.

If this event on the output port is sent by tizonia then gst-omx retrieves the params (width, height, channels etc..) and then enables the output port, allocate output buffers and populate.

A ) Usually/typically this event is emitted from tizonia from tizkernel.c::apply_slaving_behaviour, because gst-omx always set the params on the input ports (gstreamer has some parsers elements connected to the input port of gst-omx elements to they parse the stream and give the info to gst-omx). Then for audio either the sampling rate or channels or bps or nBufferSize is not the same anymore between input port and output port. (See tizpcmport.c for audio and tizvideoport.c for video), so the event OMX_EventPortSettingsChanged is sent from Tizonia to tell the client that the output port settings has changed. Which is good but I noticed this event might never be sent because the settings does not change.

Indeed the problem is that if the *initial* parameters set from tizonia/plugins/ aacdec., mp3d.c (i.e. instantiate_input_port, instantiate_output_port) are the same (i.e. input vs output) AND if these values are exactly what are in the played strea. So p_obj->pcmmode_.nSamplingRate != new_rate i never hit and the OMX_EventPortSettingsChanged is not sent. i.e. because A) detects same values.

For video, A) will mostly always detects different values because there are multiple params and the defaults are very little size 176,144 for vp8 and a real stream will not have these values.
So this is why the video case is not really a problem.

But for audio, I have a lot of files where it is 2 channels and rate 48000.
Since aacdec.c and mp3d.c defines both these values on input and output then the event is never sent and gst-omx is stuck forever.

1) Is gst-omx correct to act like this ? If not I do not see how it should behave. I would prefer not to change gst-omx to keep compat with 1.1.2 and to not make more #ifdef version

2) a simple solution is just to set nChannels =0 or nSampleRate=0 on all tizonia output ports for its audio plugins. 

Thx in advance for your feedback.
Julien

**Comment by juanrubio** [comment]
Hi Julien!,

Yes, I think I understand what is the problem you are describing. There is a small chance that this could be one of those "in 1.2 is like this and in 1.1.2 is like that", but I'm not entirely sure. So let me dig into the spec (and code) and I'll come back to you.

**Comment by juanrubio** [comment]
Hi Julien,

IMO gst-omx is wrong waiting for a PortSettingsChanged event when *no* changes are needed on the output port after parsing and obtaining the parameters of the input stream. This is true even in the context of 1.1.2.

IMO this deserves a defect in the gst-omx bug tracker. Also the spec (both 1.1.2 and 1.2) specify the default parameters that a "standard" port needs to have at instantiation. So for a PCM port (decoder's output), the default values are 2 ch and 48000 (see values in italics in 1.1.2@p324).

Changing the Tizonia component to have different "default" values would not be a good solution because it would make the aac component non-conformant in both specs.

So what do you think?

**Comment by CapOM** [comment]
You are right it looks like gst-omx is not correct. But how can it know that the component has finished parsing ? It is important to wait that because otherwise the client will allocate the buffers on the output port twice which is not optimal. Indeed one time with port default values and most likely  a second time after the SettingsChanged event.
